pkg/webhook/server.go

```diff
@@ -0,0 +1,104 @@
+package webhook
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/openservicemesh/osm/pkg/certificate"
+	"github.com/openservicemesh/osm/pkg/constants"
+	"github.com/openservicemesh/osm/pkg/metricsstore"
+)
+
+// CertRotatedFunc is a callback to perform other actions when the server's HTTPS cert gets rotated.
+type CertRotatedFunc func(cert *certificate.Certificate) error
+
+// Server is a construct to run generic HTTPS webhook servers.
+type Server struct {
+	name         string
+	cm           *certificate.Manager
+	server       *http.Server
+	onCertChange CertRotatedFunc
+
+	mu   sync.Mutex
+	cert tls.Certificate
+}
+
+// NewServer returns a new server based on the input. Run() must be called to start the server.
+func NewServer(name, namespace string, port int, cm *certificate.Manager, handlers map[string]http.HandlerFunc, onCertChange CertRotatedFunc) (*Server, error) {
+	mux := http.NewServeMux()
+
+	for path, h := range handlers {
+		mux.Handle(path, metricsstore.AddHTTPMetrics(h))
+	}
+
+	s := &Server{
+		name:         name,
+		cm:           cm,
+		onCertChange: onCertChange,
+	}
+	s.server = &http.Server{
+		Addr:    fmt.Sprintf(":%d", port),
+		Handler: mux,
+		// #nosec G402
+		TLSConfig: &tls.Config{
+			GetCertificate: func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+				s.mu.Lock()
+				defer s.mu.Unlock()
+				return &s.cert, nil
+			},
+			MinVersion: tls.VersionTLS13,
```

Comment:
Are we using this minVersion elsewhere? If any client doesn't support TLS 1.3, then the handshake will fail. I'd say 1.2 is probably safer.

Reply:
This is what was used previously, I don't think it fits in the scope of this PR to make that change